File: examples/python-assert-external/assert.py
def get_assert(output, context):
    """
    Custom function that grades an LLM output.
    """

    # You can return a simple boolean
    if "banana" not in output.lower():
        return False

    # Or a number
    if "yellow" not in output.lower():
        return 0.5  # Partial credit

    # Or return a dict with additional info
    simple_result = {
        "pass": True,  # Using "pass" as dictionary key works fine
        "score": 0.8,
        "reason": "Good response about bananas"
    }

    # Or include nested assertions...
    # Note: Both camelCase and snake_case field names are supported
    # This example uses snake_case (pass_, component_results, named_scores, tokens_used)
    result = {
        "pass_": True,  # 'pass_' only needed for dataclasses, but supported here too
        "score": 0.75,
        "reason": "Looks good to me",
        "named_scores": {
            "creativity": 0.8,
            "accuracy": 0.9
        },
        "tokens_used": {
            "total": 150,
            "prompt": 100,
            "completion": 50
        },
        "component_results": [
            {
                "pass_": "bananas" in output.lower(),
                "score": 0.5,
                "reason": "Contains banana",
                "named_scores": {
                    "Uses banana": 1.0,
                },
            },
            {
                "pass_": "yellow" in output.lower(),
                "score": 0.5,
                "reason": "Contains yellow",
                "named_scores": {
                    "Yellowish": 0.66,
                },
            },
        ],
    }

    return result

File: examples/javascript-assert-external/assert.js
module.exports = {
  customFunction: (output, context) => {
    console.log('Prompt:', context.prompt);
    console.log('Vars', context.vars.topic);

    // You can return a bool...
    // return output.toLowerCase().includes('bananas');

    // A score (where 0 = Fail)...
    // return 0.5;

    // Or an entire grading result, which can be simple...
    let result = {
      pass: output.toLowerCase().includes('bananas'),
      score: 0.5,
      reason: 'Contains banana',
    };

    // Or include nested assertions...
    result = {
      pass: true,
      score: 0.75,
      reason: 'Looks good to me',
      componentResults: [
        {
          pass: output.toLowerCase().includes('bananas'),
          score: 0.5,
          reason: 'Contains banana',
          namedScores: {
            'Uses banana': 1.0,
          },
        },
        {
          pass: output.toLowerCase().includes('yellow'),
          score: 0.5,
          reason: 'Contains yellow',
          namedScores: {
            Yellowish: 0.66,
          },
        },
      ],
    };

    return result;
  },
};
